### PR TITLE
gadget,kernel: add new kernel.{Info,Asset} struct and helpers (1/N)

### PR DESCRIPTION
**Pull request by mvo5**
This is the start of implementing the kernel assets spec as
outlined in: https://forum.snapcraft.io/t/415/46

This is really #9148 but I closed it because of an incomplete push and GH won't
let me reopen the original PR :/

### CODE REVIEW

File: gadget/edition/number.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+package edition
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+)
+
+type Number uint32
```
Comment (pedronis) [inline gadget/edition/number.go:28]: this merits a doc comment, like

// Number can hold (and unmarshal) an edition number, used in gadget.yaml and kernel.yaml to control whether
// updates should be applied.

File: kernel/kernel.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+	if err := yaml.Unmarshal(kernelYaml, &ki); err != nil {
+		return nil, fmt.Errorf("cannot parse kernel metadata: %v", err)
+	}
+
+	return &ki, nil
+}
+
+// ReadInfo reads the kernel specific metadata from meta/kernel.yaml
+// in the snap root directory.
```
Comment (bboozzoo) [inline kernel/kernel.go:54]:
```suggestion
// in the snap root directory if the file exists.
```

File: kernel/kernel.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+// InfoFromKernelYaml reads the provided kernel metadata.
+func InfoFromKernelYaml(kernelYaml []byte) (*Info, error) {
+	var ki Info
+
+	if err := yaml.Unmarshal(kernelYaml, &ki); err != nil {
+		return nil, fmt.Errorf("cannot parse kernel metadata: %v", err)
+	}
+
+	return &ki, nil
```
Comment (bboozzoo) [inline kernel/kernel.go:50]: Gadget does quite a lot of validation, perhaps we should leave a TODO to at least validate the assets for existence.

File: kernel/kernel.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+func InfoFromKernelYaml(kernelYaml []byte) (*Info, error) {
+	var ki Info
+
+	if err := yaml.Unmarshal(kernelYaml, &ki); err != nil {
+		return nil, fmt.Errorf("cannot parse kernel metadata: %v", err)
+	}
+
+	for name := range ki.Assets {
+		if !validAssetName.MatchString(name) {
+			return nil, fmt.Errorf("invalid asset name %q, please use only alphanumeric charackters", name)
```
Comment (anonymouse64) [inline kernel/kernel.go:56]:
```suggestion
			return nil, fmt.Errorf("invalid asset name %q, please use only alphanumeric characters", name)
```

**Comment by bboozzoo** [comment]
This can land, test failures are unrelated.

**Pull request merged**